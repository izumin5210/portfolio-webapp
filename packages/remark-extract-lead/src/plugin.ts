import type { Plugin } from "unified";
import type { Node, Parent } from "unist";
import { is } from "unist-util-is";

export const remarkExtractLead: Plugin<[], Parent> = () => {
  return function transformer(tree, file) {
    let [startIdx, endIdx] = [-1, -1];
    for (let i = 0; i < tree.children.length; i++) {
      const node = tree.children[i];
      if (is<YamlNode>(node, "yaml") && startIdx === -1) continue;
      if (is<HeadingNode>(node, "heading")) {
        if (node.depth === 1 && startIdx === -1) continue;
        break;
      }
      if (startIdx === -1) startIdx = i;
      endIdx = i;
    }
    file.data.lead = {
      type: "root",
      children: tree.children.slice(startIdx, endIdx + 1),
    };
  };
};

interface HeadingNode extends Parent {
  depth: number;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface YamlNode extends Node {
  type: "yaml";
}
