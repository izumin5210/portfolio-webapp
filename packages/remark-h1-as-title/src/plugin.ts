import type { Heading, Root, Text, YAML } from "mdast";
import type { Plugin } from "unified";
import { is } from "unist-util-is";

export const remarkH1AsTitle: Plugin<[], Root> = () => {
  return function transformer(tree, file) {
    for (let i = 0; i < tree.children.length; i++) {
      const node = tree.children[i];
      if (is<YAML>(node, "yaml")) continue;
      if (is<Heading>(node, "heading") && node.depth === 1) {
        if (node.children.length === 1 && is<Text>(node.children[0], "text")) {
          file.data.title = node.children[0].value;
          tree.children = [...tree.children.slice(0, i), ...tree.children.slice(i + 1)];
        } else {
          file.fail(new Error(" h1 at the beginning of the document must be text only"), node.position);
        }
        break;
      }
      file.fail(new Error("the document must start with h1"), node.position);
      break;
    }
  };
};
