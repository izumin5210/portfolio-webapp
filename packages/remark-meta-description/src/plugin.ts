import type { BlockContent, Content, Heading, List, Paragraph, Root, YAML } from "mdast";
import { toString } from "mdast-util-to-string";
import type { Plugin } from "unified";
import { is } from "unist-util-is";
import { visit } from "unist-util-visit";

export const remarkMetaDescription: Plugin<[], Root> = () => {
  return function transformer(tree, file) {
    let [startIdx, endIdx] = [-1, -1];
    for (let i = 0; i < tree.children.length; i++) {
      const node = tree.children[i];
      if (is<YAML>(node, "yaml") && startIdx === -1) continue;
      if (is<Heading>(node, "heading")) {
        if (node.depth === 1 && startIdx === -1) continue;
        break;
      }
      if (isBlockContent(node) && !is<List>(node, "list") && !is<Paragraph>(node, "paragraph")) {
        file.fail(new Error(`meta description cannot has ${node.type} content`), node.position);
      }
      for (const bc of findBlockContents(node, { except: ["paragraph"] })) {
        file.fail(new Error(`meta description cannot has ${bc.type} content in ${node.type}`), bc.position);
      }

      if (startIdx === -1) startIdx = i;
      endIdx = i;
    }

    const descMdNodes = tree.children.slice(startIdx, endIdx + 1);
    for (let i = 0; i < descMdNodes.length; i++) {
      const node = descMdNodes[i];
      if (is<List>(node, "list")) {
        descMdNodes[i] = {
          type: "paragraph",
          children: node.children.map((li) => ({ type: "text", value: ` Â· ${toString(li)}; ` })),
        };
      }
    }

    file.data.metaDescription = toString({ type: "root", children: descMdNodes });
  };
};

function findBlockContents(content: Content, opts: { except: BlockContent["type"][] }): BlockContent[] {
  const blockContents: BlockContent[] = [];
  visit(content, (child) => {
    if (content === child) return;
    if (isBlockContent(child) && !opts.except.includes(child.type)) blockContents.push(child);
  });
  return blockContents;
}

function isBlockContent(content: Content): content is BlockContent {
  switch (content.type) {
    case "paragraph":
    case "heading":
    case "blockquote":
    case "thematicBreak":
    case "list":
    case "table":
    case "html":
    case "code": {
      const _blockContentCheck: BlockContent = content;
      return true;
    }
    default: {
      const _nonBlockContentCheck: Exclude<Content, BlockContent> = content;
      return false;
    }
  }
}
