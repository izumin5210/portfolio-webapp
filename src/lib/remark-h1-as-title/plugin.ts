import { is } from "unist-util-is";
import type { Plugin } from "unified";
import type { Node, Parent } from "unist";

export const remarkH1AsTitlePlugin: Plugin<[], Parent> = () => {
  return function transformer(tree, file) {
    for (let i = 0; i < tree.children.length; i++) {
      const node = tree.children[i];
      if (node.type === "yaml") continue;
      if (is(node, "heading") && (node as { depth?: number }).depth === 1) {
        const h1 = node as Parent;
        if (h1.children.length === 1 && h1.children[0].type === "text") {
          file.data.title = (h1.children[0] as Node & { value: string }).value;
          tree.children = [...tree.children.slice(0, i), ...tree.children.slice(i + 1)];
        } else {
          file.fail(new Error(" h1 at the beginning of the document must be text only"), (node as Node).position);
        }
        break;
      }
      file.fail(new Error("the document must start with h1"), (node as Node).position);
      break;
    }
  };
};
