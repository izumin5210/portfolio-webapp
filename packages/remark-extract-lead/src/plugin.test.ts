import remarkFrontmatter from "remark-frontmatter";
import remarkParse from "remark-parse";
import remarkStringify from "remark-stringify";
import remarkRehype from "remark-rehype";
import rehypeStringify from "rehype-stringify";
import { unified } from "unified";
import { remarkExtractLead } from "./plugin.js";
import { Node } from "unist";

const processor = unified().use(remarkParse).use(remarkFrontmatter).use(remarkExtractLead).use(remarkStringify);

const mdAstToHtmlString = (root: Node) => {
  const rehypeTree = unified()
    .use(remarkRehype)
    .runSync(root as any);
  return unified()
    .use(rehypeStringify)
    .stringify(rehypeTree as any);
};

it("extracts lead contents", () => {
  const file = processor.processSync(`
foobarbaz

- item 1
- [item 2](./item2)
- [item 3](./item3)

# heading 2
foobabaz
`);
  expect(mdAstToHtmlString(file.data.lead as any)).toMatchSnapshot();
});

it("returns all contents when there are no headings", () => {
  const file = processor.processSync(`
foobarbaz

- item 1
- [item 2](./item2)
- [item 3](./item3)
`);
  expect(mdAstToHtmlString(file.data.lead as any)).toMatchSnapshot();
});

it("ignores leading h1", () => {
  const file = processor.processSync(`
# heading 1
foobarbaz

# heading 2
quxquux
`);
  expect(mdAstToHtmlString(file.data.lead as any)).toMatchSnapshot();
});

it("ignores yaml frontmatter", () => {
  const file = processor.processSync(`---
foo: bar
---

# heading 1
foobarbaz

# heading 2
quxquux
`);
  expect(mdAstToHtmlString(file.data.lead as any)).toMatchSnapshot();
});
