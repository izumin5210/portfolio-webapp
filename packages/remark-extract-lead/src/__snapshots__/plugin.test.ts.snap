// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extracts lead contents 1`] = `
"<p>foobarbaz</p>
<ul>
<li>item 1</li>
<li><a href=\\"./item2\\">item 2</a></li>
<li><a href=\\"./item3\\">item 3</a></li>
</ul>"
`;

exports[`ignores leading h1 1`] = `"<p>foobarbaz</p>"`;

exports[`ignores yaml frontmatter 1`] = `"<p>foobarbaz</p>"`;

exports[`returns all contents when there are no headings 1`] = `
"<p>foobarbaz</p>
<ul>
<li>item 1</li>
<li><a href=\\"./item2\\">item 2</a></li>
<li><a href=\\"./item3\\">item 3</a></li>
</ul>"
`;
